query-node/mappings/src/workingGroup.ts

```diff
@@ -0,0 +1,180 @@
+  // read event data
+  const {applicationIdToWorkerIdMap} = new GatewayWorkingGroup.OpeningFilledEvent(event).data
+
+  // call generic processing
+  await workingGroup_OpeningFilled(db, StorageProviderType.GATEWAY, applicationIdToWorkerIdMap)
+}
+
+export async function gatewayWorkingGroup_WorkerStorageUpdated(db: DatabaseManager, event: SubstrateEvent): Promise<void> {
+  // read event data
+  const {workerId: storageProviderId, bytes: newMetadata} = new GatewayWorkingGroup.WorkerStorageUpdatedEvent(event).data
+
+  // call generic processing
+  await workingGroup_WorkerStorageUpdated(db, storageProviderId, newMetadata)
+}
+
+export async function gatewayWorkingGroup_TerminatedWorker(db: DatabaseManager, event: SubstrateEvent): Promise<void> {
+  // read event data
+  const {workerId: storageProviderId} = new StorageWorkingGroup.TerminatedWorkerEvent(event).data
+
+  // call generic processing
+  await workingGroup_TerminatedWorker(db, storageProviderId)
+}
+
+export async function gatewayWorkingGroup_WorkerExited(db: DatabaseManager, event: SubstrateEvent): Promise<void> {
+  // read event data
+  const {workerId: storageProviderId} = new StorageWorkingGroup.WorkerExitedEvent(event).data
+
+  // call generic processing
+  await workingGroup_WorkerExited(db, storageProviderId)
+}
+
+/////////////////// Generic working group processing ///////////////////////////
+
+export async function workingGroup_OpeningFilled(
+  db: DatabaseManager,
+  storageProviderType: StorageProviderType,
+  applicationIdToWorkerIdMap: ApplicationIdToWorkerIdMap
+): Promise<void> {
+  const storageProviderIds = [...applicationIdToWorkerIdMap.values()]
+
+  for (const storageProviderId of storageProviderIds) {
+    // load storage provider
+    const storageProvider = await db.get(StorageProvider, { where: { id: storageProviderId.toString() } as FindConditions<StorageProvider> })
+
+    // reactivate storage provider if it already exists
+    if (storageProvider) {
+      await reactivateStorageProvider(db, storageProvider)
+      continue
+    }
+
+    // create new storage provider
+    const newStorageProvider = new StorageProvider({
+      id: storageProviderId.toString(),
+      type: storageProviderType,
+      isActive: true,
+    })
+
+    await db.save<StorageProvider>(newStorageProvider)
+  }
+
+  // emit log event
+  logger.info("Storage provider has beed created", {ids: storageProviderIds.map(item => item.toString())})
+}
+
+export async function workingGroup_WorkerStorageUpdated(db: DatabaseManager, storageProviderId: WorkerId, newMetadata: Bytes): Promise<void> {
+  // load storage provider
+  const storageProvider = await db.get(StorageProvider, { where: { id: storageProviderId.toString() } as FindConditions<StorageProvider> })
+
+  // ensure storageProvider exists
+  if (!storageProvider) {
+    return inconsistentState('Non-existing storage provider update requested', storageProviderId)
```

Comment:
message can be improved and generalized to a worker not storage provider

Reply:
Messages has been improved.